lib/documents/document_session.rb

```diff
@@ -0,0 +1,457 @@
+      check_result = check_association_and_change_vectore_before_store(document, id, change_vector)
+      document = check_result[:document]
+      is_new_document = check_result[:is_new]
+
+      if is_new_document
+        original_metadata = DeepClone.clone(document.instance_variable_get('@metadata'))
+        document = prepare_document_id_before_store(document, id)
+        id = conventions.get_id_from_document(document)
+
+        @defer_commands.each {|command| raise InvalidOperationException,
+          "Can't store document, there is a deferred command registered "\
+          "for this document in the session. Document id: #{id}" if
+          id == command.document_id
+        }
+
+        raise InvalidOperationException,
+          "Can't store object, it was already deleted in this "\
+          "session. Document id: #{id}" if
+          @deleted_documents.include?(document)
+
+        on_document_fetched({
+          :document => document,
+          :metadata => document.instance_variable_get('@metadata'),
+          :original_metadata => original_metadata,
+          :raw_entity => conventions.convert_to_raw_entity(document)
+        })
+      end
+
+      document
+    end
+
+    def save_changes
+      changes = SaveChangesData.new(@defer_commands.to_a, @defer_commands.size)
+
+      @defer_commands.clear
+      prepare_delete_commands(changes)
+      prepare_update_commands(changes)
+
+      if !changes.commands_count
+        return nil
+      end
+
+      results = @request_executor.execute(changes.create_batch_command)
+
+      if !results
+        raise InvalidOperationException.new, "Cannot call Save Changes after the document store was disposed."
+      end
+
+      process_batch_command_results(results, changes)
+    end
+
+    protected
+    def attach_query(query)
+      if @attached_queries.key?(query)
+        raise InvalidOperationException, 'Query is already attached to session'
+      end
+
+      query.on(RavenServerEvent::EVENT_DOCUMENTS_QUERIED) {
+        increment_requests_count
+      }
+
+      query.on(RavenServerEvent::EVENT_DOCUMENT_FETCHED) {|conversion_result|
+        on_document_fetched(conversion_result)
+      }
+
+      query.on(RavenServerEvent::EVENT_INCLUDES_FETCHED) {|includes|
+        on_includes_fetched(includes)
+      }
+
+      @attached_queries[query] = true
+    end
+
+    def increment_requests_count
+      max_requests = DocumentConventions::MaxNumberOfRequestPerSession
+
+      @number_of_requests_in_session = @number_of_requests_in_session + 1
+
+      raise InvalidOperationException,
+          "The maximum number of requests (#{max_requests}) allowed for this session has been reached. Raven limits the number "\
+"of remote calls that a session is allowed to make as an early warning system. Sessions are expected to "\
+"be short lived, and Raven provides facilities like batch saves (call saveChanges() only once) "\
```

Comment:
Please review this error message and apply fixed to things like `saveChanges` --> `save_changes`